components/brave_rewards/browser/database/database_recurring_tip.cc

```diff
@@ -0,0 +1,181 @@
+#include "sql/statement.h"
+#include "sql/transaction.h"
+
+namespace brave_rewards {
+
+namespace {
+  // TODO(nejczdovc): rename to recurring_tip
+  const char* table_name_ = "recurring_donation";
+  const int minimum_version_ = 2;
+}  // namespace
+
+DatabaseRecurringTip::DatabaseRecurringTip(
+    int current_db_version) :
+    DatabaseTable(current_db_version) {
+}
+
+DatabaseRecurringTip::~DatabaseRecurringTip() = default;
+
+bool DatabaseRecurringTip::Init(sql::Database* db) {
+  if (GetCurrentDBVersion() < minimum_version_) {
+    return true;
+  }
+
+  sql::Transaction transaction(db);
+  if (!transaction.Begin()) {
+    return false;
+  }
+
+  bool success = CreateTable(db);
+  if (!success) {
+    return false;
+  }
+
+  success = CreateIndex(db);
+  if (!success) {
+    return false;
+  }
+
+  return transaction.Commit();
+}
+
+bool DatabaseRecurringTip::CreateTable(sql::Database* db) {
+  if (db->DoesTableExist(table_name_)) {
+    return true;
+  }
+
+  return CreateTableV2(db);
+}
+
+bool DatabaseRecurringTip::CreateTableV2(sql::Database* db) {
+  const std::string query = base::StringPrintf(
+      "CREATE TABLE %s ("
+        "publisher_id LONGVARCHAR NOT NULL PRIMARY KEY UNIQUE,"
+        "amount DOUBLE DEFAULT 0 NOT NULL,"
+        "added_date INTEGER DEFAULT 0 NOT NULL,"
+        "CONSTRAINT fk_%s_publisher_id"
+        "    FOREIGN KEY (publisher_id)"
+        "    REFERENCES publisher_info (publisher_id)"
+        "    ON DELETE CASCADE"
+      ")",
+      table_name_,
+      table_name_);
+
+  return db->Execute(query.c_str());
+}
+
+bool DatabaseRecurringTip::CreateIndex(sql::Database* db) {
+  return CreateIndexV2(db);
+}
+
+bool DatabaseRecurringTip::CreateIndexV2(sql::Database* db) {
+  return this->InsertIndex(db, table_name_, "publisher_id");
+}
+
+bool DatabaseRecurringTip::Migrate(sql::Database* db, const int target) {
+  switch (target) {
+    case 2: {
+      return MigrateToV2(db);
+    }
+    default: {
+      NOTREACHED();
+      return false;
+    }
+  }
+}
+
+bool DatabaseRecurringTip::MigrateToV2(sql::Database* db) {
+  if (!CreateTableV2(db)) {
+    return false;
+  }
+
+  if (!CreateIndexV2(db)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool DatabaseRecurringTip::InsertOrUpdate(
+    sql::Database* db,
+    ledger::RecurringTipPtr info) {
+  if (!info || info->publisher_key.empty()) {
+    return false;
+  }
+
+  const std::string query = base::StringPrintf(
+      "INSERT OR REPLACE INTO %s "
+      "(publisher_id, amount, added_date) "
+      "VALUES (?, ?, ?)",
+      table_name_);
+
+  sql::Statement statement(
+      db->GetCachedStatement(SQL_FROM_HERE, query.c_str()));
+
+  statement.BindString(0, info->publisher_key);
+  statement.BindDouble(1, info->amount);
+  statement.BindInt64(2, info->created_at);
+
+  return statement.Run();
+}
+void DatabaseRecurringTip::GetAllRecords(
+    sql::Database* db,
+    ledger::PublisherInfoList* list) {
+  if (!list) {
```

Comment:
`DCHECK(list)` to identify potential issues when running debug builds

Reply:
`DCHECK(db)` and `if (!db) {...` to identify potential issues when running debug builds

Reply:
db covered in https://github.com/brave/brave-browser/issues/7145

Reply:
list done